File: backend/app/core/entitlements.py
"""Controle d acces par feature et plan SOMA."""
from datetime import datetime, timezone
from fastapi import Depends, HTTPException, status

from app.core.plans import PlanCode
from app.core.features import FeatureCode
from app.core.deps import get_current_user
from app.models.user import User


# Matrice feature -> plan minimum requis
_FEATURE_MIN_PLAN: dict[FeatureCode, PlanCode] = {
    FeatureCode.BASIC_DASHBOARD: PlanCode.FREE,
    FeatureCode.BASIC_HEALTH_METRICS: PlanCode.FREE,
    FeatureCode.LOCAL_AI_TIPS: PlanCode.FREE,
    FeatureCode.AI_COACH: PlanCode.AI,
    FeatureCode.DAILY_BRIEFING: PlanCode.AI,
    FeatureCode.ADVANCED_INSIGHTS: PlanCode.AI,
    FeatureCode.PDF_REPORTS: PlanCode.AI,
    FeatureCode.ANOMALY_DETECTION: PlanCode.AI,
    FeatureCode.BIOLOGICAL_AGE: PlanCode.AI,
    FeatureCode.READINESS_SCORE: PlanCode.PERFORMANCE,
    FeatureCode.INJURY_PREDICTION: PlanCode.PERFORMANCE,
    FeatureCode.BIOMECHANICS_VISION: PlanCode.PERFORMANCE,
    FeatureCode.ADVANCED_VO2MAX: PlanCode.PERFORMANCE,
    FeatureCode.TRAINING_LOAD: PlanCode.PERFORMANCE,
}

# Matrice plan -> set de features incluses
PLAN_FEATURES: dict[PlanCode, set[FeatureCode]] = {
    PlanCode.FREE: {
        FeatureCode.BASIC_DASHBOARD,
        FeatureCode.BASIC_HEALTH_METRICS,
        FeatureCode.LOCAL_AI_TIPS,
    },
    PlanCode.AI: {
        FeatureCode.BASIC_DASHBOARD,
        FeatureCode.BASIC_HEALTH_METRICS,
        FeatureCode.LOCAL_AI_TIPS,
        FeatureCode.AI_COACH,
        FeatureCode.DAILY_BRIEFING,
        FeatureCode.ADVANCED_INSIGHTS,
        FeatureCode.PDF_REPORTS,
        FeatureCode.ANOMALY_DETECTION,
        FeatureCode.BIOLOGICAL_AGE,
    },
    PlanCode.PERFORMANCE: {
        FeatureCode.BASIC_DASHBOARD,
        FeatureCode.BASIC_HEALTH_METRICS,
        FeatureCode.LOCAL_AI_TIPS,
        FeatureCode.AI_COACH,
        FeatureCode.DAILY_BRIEFING,
        FeatureCode.ADVANCED_INSIGHTS,
        FeatureCode.PDF_REPORTS,
        FeatureCode.ANOMALY_DETECTION,
        FeatureCode.BIOLOGICAL_AGE,
        FeatureCode.READINESS_SCORE,
        FeatureCode.INJURY_PREDICTION,
        FeatureCode.BIOMECHANICS_VISION,
        FeatureCode.ADVANCED_VO2MAX,
        FeatureCode.TRAINING_LOAD,
    },
}


def get_effective_plan(user: User) -> PlanCode:
    """
    Source de verite unique pour le plan effectif d un utilisateur.
    Regles:
      - plan_status != active -> FREE (carte expiree, annulation, paiement echoue)
      - plan_expires_at depasse -> FREE (meme si plan_code pas encore nettoye)
      - sinon: plan_code de la DB
    """
    if user.plan_status != "active":
        return PlanCode.FREE
    if user.plan_expires_at and user.plan_expires_at < datetime.now(timezone.utc):
        return PlanCode.FREE
    # Trial check: si trial expire, retour au free
    if user.trial_ends_at and user.trial_ends_at < datetime.now(timezone.utc):
        # Si plan_code = free en trial, on reste free
        if user.plan_code == PlanCode.FREE:
            return PlanCode.FREE
    try:
        return PlanCode(user.plan_code)
    except ValueError:
        return PlanCode.FREE


def user_has_feature(user: User, feature: FeatureCode) -> bool:
    """Retourne True si l utilisateur a acces a la feature selon son plan effectif."""
    effective = get_effective_plan(user)
    return feature in PLAN_FEATURES.get(effective, set())


def get_user_features(user: User) -> list[str]:
    """Retourne la liste des feature codes accessibles pour l utilisateur."""
    effective = get_effective_plan(user)
    return [f.value for f in PLAN_FEATURES.get(effective, set())]


def require_feature(feature: FeatureCode):
    """FastAPI dependency: leve HTTPException 403 si la feature n est pas dans le plan effectif."""
    def checker(current_user: User = Depends(get_current_user)) -> User:
        if not user_has_feature(current_user, feature):
            effective = get_effective_plan(current_user)
            min_plan = _FEATURE_MIN_PLAN.get(feature, PlanCode.AI)
            raise HTTPException(
                status_code=status.HTTP_403_FORBIDDEN,
                detail={
                    "error": "feature_not_available",
                    "feature": feature.value,
                    "required_plan": min_plan.value,
                    "current_plan": current_user.plan_code,
                    "effective_plan": effective.value,
                    "plan_status": current_user.plan_status,
                    "message": (
                        f"La fonctionnalite '{feature.value}' necessite le plan "
                        f"'{min_plan.value}'. Plan effectif: '{effective.value}'."
                    ),
                },
            )
        return current_user
    return checker


def require_superuser(current_user: User = Depends(get_current_user)) -> User:
    """FastAPI dependency: reserve aux superusers (admin)."""
    if not getattr(current_user, "is_superuser", False):
        raise HTTPException(
            status_code=status.HTTP_403_FORBIDDEN,
            detail="Admin access required",
        )
    return current_user

File: desktop/src/lib/entitlements.ts
// Feature codes matching backend FeatureCode enum exactly
export const FeatureCode = {
  // Free tier
  BASIC_DASHBOARD: 'basic_dashboard',
  BASIC_HEALTH_METRICS: 'basic_health_metrics',
  LOCAL_AI_TIPS: 'local_ai_tips',

  // AI tier
  AI_COACH: 'ai_coach',
  DAILY_BRIEFING: 'daily_briefing',
  ADVANCED_INSIGHTS: 'advanced_insights',
  PDF_REPORTS: 'pdf_reports',
  ANOMALY_DETECTION: 'anomaly_detection',
  BIOLOGICAL_AGE: 'biological_age',

  // Performance tier
  READINESS_SCORE: 'readiness_score',
  INJURY_PREDICTION: 'injury_prediction',
  BIOMECHANICS_VISION: 'biomechanics_vision',
  ADVANCED_VO2MAX: 'advanced_vo2max',
  TRAINING_LOAD: 'training_load',

  // Coach Module (human coaching platform)
  COACH_MODULE: 'coach_module',
  COACH_ADVANCED_ANALYSIS: 'coach_advanced_analysis',
  COACH_REPORTS: 'coach_reports',
} as const;

export type FeatureCodeType = typeof FeatureCode[keyof typeof FeatureCode];

export interface EntitlementsData {
  plan_code: string;
  plan_status: string;
  features: string[];
  plan_expires_at?: string | null;
  trial_ends_at?: string | null;
  is_trial: boolean;
  is_expired: boolean;
}

export function hasFeature(
  data: EntitlementsData | undefined,
  feature: string
): boolean {
  if (!data) return false;
  return data.features.includes(feature);
}

export function getRequiredPlan(feature: string): 'ai' | 'performance' {
  const performanceFeatures = new Set<string>([
    FeatureCode.READINESS_SCORE,
    FeatureCode.INJURY_PREDICTION,
    FeatureCode.BIOMECHANICS_VISION,
    FeatureCode.ADVANCED_VO2MAX,
    FeatureCode.TRAINING_LOAD,
    FeatureCode.COACH_ADVANCED_ANALYSIS,
  ]);
  return performanceFeatures.has(feature) ? 'performance' : 'ai';
}

export function getPlanDisplayName(planCode: string): string {
  switch (planCode) {
    case 'ai': return 'SOMA AI';
    case 'performance': return 'SOMA Performance';
    default: return 'SOMA Free';
  }
}

export function getPlanBadgeColor(planCode: string): string {
  switch (planCode) {
    case 'ai': return 'bg-blue-500/20 text-blue-400 border-blue-500/30';
    case 'performance': return 'bg-purple-500/20 text-purple-400 border-purple-500/30';
    default: return 'bg-gray-500/20 text-gray-400 border-gray-500/30';
  }
}
